honda bosch longitudinal safety

gregjhogan opened the pull request:
TODO:
- [x] brake limit should be -350 when moving and -400 when stopped (or can we use -350 for standstill?) -350 seems fine
- [x] standstill bit dangerous to send when moving? seems fine, probably smooths out braking
- [x] openpilot should set `long_controls_allowed = false` when running stock bosch? (no longer exists)
- [x] always allow sending ACC messages? no
- [x] tests
- [x] split code out into new safety mode

geohot commented:
How do you enable this? Do you need to unplug the radar?

gregjhogan commented:
There are several options:
- unplug radar
- enter diagnostic session and send communication control command to disable tx/rx, all communication, suppress response
  (address `0x18DAB0F1` send `\x02\x10\x03` then `\x03\x28\x83\x03`)
  something is still sent indicating a fault state, but since non-fault is dominant you can override by sending everything is OK (if you send it fast enough)
- enter programming session

rbiasini commented on board/safety/safety_honda.h:
```diff
@@ -144,14 +162,24 @@ static int honda_tx_hook(CAN_FIFOMailBox_TypeDef *to_send) {
+    int gas = (GET_BYTE(to_send, 0) << 8) | GET_BYTE(to_send, 1);
+    gas = to_signed(gas, 16);
+    if (!current_controls_allowed || !long_controls_allowed) {
+      if (gas != -30000) {
```
I would parametrize these values

gregjhogan replied:
fixed

rbiasini commented on board/safety/safety_honda.h:
```diff
@@ -133,6 +135,22 @@ static int honda_tx_hook(CAN_FIFOMailBox_TypeDef *to_send) {
     }
   }
+  // BRAKE: safety check (bosch)
+  if (addr == 0x1DF) {
```
specify bus, depending on panda type

gregjhogan replied:
fixed

rbiasini commented on board/safety/safety_honda.h:
```diff
@@ -144,14 +162,24 @@ static int honda_tx_hook(CAN_FIFOMailBox_TypeDef *to_send) {
     }
   }
+  // GAS: safety check (bosch)
+  if (addr == 0x1DF) {
```
can be merged with brake check. Is the same message

gregjhogan replied:
fixed

gregjhogan commented on board/safety/safety_honda.h:
```diff
@@ -7,9 +7,13 @@
+// TODO: only allow TX messages on bus 0 if OP longitudinal control?
+const AddrBus HONDA_BG_TX_MSGS[] = {{0xE4, 0}, {0xE4, 2}, {0x1DF, 0}, {0x1EF, 0}, {0x296, 0}, {0x30C, 0}, {0x33D, 0}, {0x33D, 2}, {0x39F, 0}};  // Bosch Giraffe
+// TODO: only allow TX messages on bus 1 if OP longitudinal control?
+const AddrBus HONDA_BH_TX_MSGS[] = {{0xE4, 0}, {0xE4, 1}, {0x1DF, 1}, {0x1EF, 1}, {0x296, 1}, {0x30C, 1}, {0x33D, 0}, {0x33D, 1}, {0x39F, 1}};  // Bosch Harness
```
@rbiasini is it OK to always allow steering/gas/brake messages on powertrain bus?

gregjhogan commented:
I think this is ready to merge now.  Drove this branch with OP master without any issues.

rbiasini commented:
I know we discussed this the other night: I think it's fine (from a safety standpoint) to allow a message to be sent on both buses, but since this use case (long control for bosch) is for now way less frequent than the LaC only, I recommend the following:

* Not worth supporting the giraffe config IMO. It will probably be deprecated at a certain point, not worth adding features on top of it.
* create a separate safety mode instead of overloading hondaBoschHarness. It could be called hondaBoschNoRadar or similar, and set a variable (`honda_bosch_no_radar=true` ?) so that in the tx_hook you can decide which checks apply. 
* Have a separate safety test class, which inherits from honda_bosch_harness and runs extra checks for gas and brakes.

Sorry, I would have recommended these changes during the first review, but the safety architecture changed a bit recently and this is the way I would do it now.

rbiasini commented on board/safety/safety_honda.h:
```diff
@@ -7,9 +7,12 @@
-const AddrBus HONDA_BG_TX_MSGS[] = {{0xE4, 2}, {0x296, 0}, {0x33D, 2}};  // Bosch Giraffe
-const AddrBus HONDA_BH_TX_MSGS[] = {{0xE4, 0}, {0x296, 1}, {0x33D, 0}};  // Bosch Harness
+const AddrBus HONDA_BG_TX_MSGS[] = {{0xE4, 0}, {0xE4, 2}, {0x1DF, 0}, {0x1EF, 0}, {0x296, 0}, {0x30C, 0}, {0x33D, 0}, {0x33D, 2}, {0x39F, 0}};  // Bosch Giraffe
+const AddrBus HONDA_BH_TX_MSGS[] = {{0xE4, 0}, {0xE4, 1}, {0x1DF, 1}, {0x1EF, 1}, {0x296, 1}, {0x30C, 1}, {0x33D, 0}, {0x33D, 1}, {0x39F, 1}};  // Bosch Harness
```
hesitant to allow gas and brake messages to be sent for the normal case where only lateral control is being executed. Better to define a separate TX_MSGS list like:
```
HONDA_BG_TX_MSGS[] = {{0xE4, 1}, {0x1DF, 1}, ...};  // Bosch Harness with radar disconnected
```
And use it in its own safety mode.

gregjhogan commented:
updated to enable bosch longitudinal control messages based on a param passed when setting safety mode, should be good to merge once build is green

@pd0wm please review - TX whitelist is now broke out (gas/brake only allowed when explicitly enabled)
@quillford please review - re-organized tests in an attempt to minimize code duplication

also noticed that there was no checksum check on alt brake message, so I fixed that, too

quillford commented on board/safety.h:
```diff
@@ -249,6 +249,10 @@ int to_signed(int d, int bits) {
   return d_signed;
 }
 
+bool get_param_flag(int16_t param, uint16_t mask) {
```
I don't think this really needs a function, but if we keep it, it should probably be more generic like `get_flag` so it can be used for the unsafe modes too.

gregjhogan replied:
due to misra checks this requires a cast from `int16_t` to `uint16_t` (see latest commit) and the fact that it uses int16 in general makes it somewhat specific to safety mode param.  Should I move it and re-name it to `get_int16_flag()` or something?

adeebshihadeh replied:
hmm, then maybe just make it a macro?

gregjhogan replied:
good idea, the macro can abstract away the type

quillford commented on board/safety/safety_honda.h:
```diff
@@ -211,8 +226,9 @@ static int honda_tx_hook(CAN_FIFOMailBox_TypeDef *to_send) {
 
-  // BRAKE: safety check
+  // BRAKE: safety check (nidec)
   if ((addr == 0x1FA) && (bus == 0)) {
```
bus == bus_pt?

gregjhogan replied:
yeah, I can update

adeebshihadeh commented on board/safety/safety_honda.h:
```diff
@@ -228,6 +244,31 @@ static int honda_tx_hook(CAN_FIFOMailBox_TypeDef *to_send) {
 
+  // BRAKE/GAS: safety check (bosch)
+  if ((addr == 0x1DF) && (bus == bus_pt)) {
+    int accel = (GET_BYTE(to_send, 3) << 3) + ((GET_BYTE(to_send, 4) >> 5) & 0x7);
```
or instead of plus

gregjhogan replied:
done

adeebshihadeh commented:
LGTM, nice catch on the missing checksum check

gregjhogan commented:
changes complete, good to merge?

gregjhogan commented:
@pd0wm rebase complete and ready to merge if no one sees any additional issues

gregjhogan commented:
removing 0x1BE brake message RX check because adding it means that message is required (breaks cars that do not use the lat brake message)
